### PR TITLE
core/ofi_signal: Allow retries in fd_signal_reset

### PR DESCRIPTION
**Pull request by dshinaberry**
Allow upto 256 spinning retries in fd_signal_reset. On Windows the atomic int32_t can be set prior to the byte written to the socket being available to be read.

on-behalf-of: @cmru-ps <dshinaberry@mru.medical.canon>
Signed-off-by: Derek Shinaberry <dshinaberry@mru.medical.canon>

### CODE REVIEW

File: include/ofi_signal.h [hunk 1]
```diff
@@ -155,10 +156,11 @@ static inline void fd_signal_reset(struct fd_signal *signal)
...
 				ofi_atomic_set32(&signal->state, OFI_SIGNAL_SET);
 
-				/* Avoid spinning forever in this highly
-				 * unlikely code path.
-				 */
-				break;
+				if (++retries > 256)
+				{
+					/* Avoid spinning forever. */
+					break;
```
Comment (j-xiong) [inline include/ofi_signal.h:162]: reset retries to 0 before the break?

**Review by j-xiong — COMMENTED** [review]
The title of the commit message is too long.

File: include/ofi_signal.h (hunk 1 above)
Comment (dshinaberry) [reply include/ofi_signal.h:162]: There is no code after the loop that the break will terminate, so I'm not sure I understand what benefit setting retries to 0 has. I feel like maybe I'm not seeing something that ought to be obvious?

Comment (j-xiong) [reply include/ofi_signal.h:162]: never mind. I was thinking about static variable across multiple calls, which is not the case here,

**Comment by shefty** [comment]
Can you expand in the commit message what the problem is that the change is trying to solve?  Are we dealing with a more complicated race condition here that requires re-working what this code is trying to do?

**Comment by dshinaberry** [comment]
I don't have visibility into the failure for the continuous-integration/jenkins/pr-head failure. The link appears to lead to an internal server at intel.com. Can someone with visibility provide some guidance as to what the issue is with the commit?

**Comment by j-xiong** [comment]
Please rebase to solve the CI failure and add "core:" or "include:" prefix to the commit title.

**Comment by j-xiong** [comment]
Also squash the two commits.

**Comment by aakefbs** [comment]
I'm just worried that these retries might mask other issues and the fixed retry value might be system dependent.

**Comment by shefty** [comment]
I think the code as structured won't work because of the race you're pointing out.  We can't maintain a state in user space and expect that to correspond with the state of the kernel socket.  We may need to "revert" the code closer to what it was prior to commit 5b635667e04917427aa5a0bc33b5936e2aa00149.  Instead trying to align states, use a counter to track how many bytes have been written to the socket with a lock to protect that value.  We're creating some very complex code here to set/reset a signal just to avoid a lock.

**Comment by shefty** [comment]
See if #7544 resolves the issue for you.  It results in easier code to follow, replaces atomics with mutex, and includes a questionably phrased commit message that the author didn't spend much time thinking about which is why it's so long

**Comment by dshinaberry** [comment]
Well, as fate would have it, later code changes appear to have changed the timing dynamics in my particular setup and I am no longer triggering the problematic code path like I was when I first encountered the issue.

I suppose I could write something simple to purposely attempt to trigger that code path and see how #7544 reacts. But I'd rather move on to the rest of the changes I have in the queue and push this to the back burner for now.

An observation about #7544, I have no idea what guarantees Windows makes regarding the completion of a socket send and the availability of that data via a socket recv. I am assuming that the locking does not really impact the data availability and only adds a small finite delay between the producer and consumer threads. If the data is not immediately available, then the call to ofi_read_socket could fail, causing the assert to fail. If you remove the assert, then it becomes a potentially infinite loop. This then reduces to essentially the same scenario in the code that I submitted but without an arbitrarily chosen limit to the number of loop iterations attempted.

**Comment by shefty** [comment]
With #7544, we could probably keep the socket as blocking and remove the loop.  The result should be the same.  I can't see in the git logs why the read socket was set to nonblocking to start with.  We could also call select/poll on the socket until it is readable (which it should be almost immediately).  That would let us set a timeout rather than using a counter.

I will see if there's any issue keeping the socket blocking.  It might depend on where else the fd is used.

**Comment by shefty** [comment]
@dshinaberry  - I don't need you to reproduce the problem, just to verify that the updated #7544 changes don't break anything in your setup.  If the changes there pass all other CI I plan on merging it and pulling them into v1.15.

**Pull request closed without merging**